core/src/main/java/org/eclipse/hono/util/EncodedPassword.java

```diff
@@ -0,0 +1,114 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Red Hat - initial creation
+ *******************************************************************************/
+package org.eclipse.hono.util;
+
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+import java.util.Base64;
+
+/**
+ * Helper class for parsing and encoding passwords.
+ * <p>
+ * The encoded password have three main elements:
+ * <ul>
+ *     <li><em>Hash function</em> that are used to encode the password</li>
+ *     <li><em>Salt</em> which is optionally used for encoding</li>
+ *     <li><em>Password</em> we want to encode</li>
+ * </ul>
+ * <p>
+ * Encoded password is returned in the {salt}password format. It can also be represented in {hashFunction}{salt}password format.
+ */
+public class EncodedPassword {
+    private static final String PREFIX = "{";
+    private static final String SUFFIX = "}";
+
+    byte[] salt;
+    String password;
+    String hashFunction =  CredentialsConstants.DEFAULT_HASH_FUNCTION;
+
+    /**
+     * Creates a new instance using hash function, Base64 encoded salt and password.
+     *
+     * @param hashFunction Hash function used to encode this password
+     * @param encodedSalt Base64 representation of the salt (if any)
+     * @param password Password to be encoded
+     */
+    public EncodedPassword(final String hashFunction, final String encodedSalt, final String password) {
+        this.hashFunction = hashFunction;
+        if (encodedSalt != null) {
+            this.salt = Base64.getDecoder().decode(encodedSalt);
+        }
+        this.password = password;
+    }
+
+    /**
+     * Creates a new instance using hash function, salt in the form of byte array and password.
+     *
+     * @param hashFunction Hash function used to encode this password
+     * @param salt Salt in the form of byte array
+     * @param password Password to be encoded
+     */
+    public EncodedPassword(final String hashFunction, final byte[] salt, final String password) {
+        this.salt = salt;
+        this.password = password;
+        this.hashFunction = hashFunction;
+    }
+
+    /**
+     * Creates a new instance from an encoded password in the {salt}password format.
+     *
+     * @param encodedPassword Encoded password in the {salt}password format
+     */
+    public EncodedPassword(final String encodedPassword) {
+        final int start = encodedPassword.indexOf(PREFIX);
+        password = encodedPassword;
+        if (start == 0) {
+            final int end = encodedPassword.indexOf(SUFFIX, start);
+            if (end > 0) {
+                salt = Base64.getDecoder().decode(encodedPassword.substring(start + 1, end));
+                password = encodedPassword.substring(end + 1);
+            }
+        }
+    }
+
+    /**
+     * Encode the password based on the provided hash function and salt.
+     *
+     * @return Encoded password
+     */
+    String encode() {
```

Comment:
why is this method here? isn't this supposed to be the job of `PasswordEncoder`?

Reply:
PasswordEncoder.encode() expects password in {salt}password format, so this helper function formats it. It also finds an appropriate encoder.